### PR TITLE
security: support multiple CN for TLS connections

### LINKED ISSUE #5134: Multiple CN support for TLS connections
## Bug Report

I want to use different CN for clients and TiKV nodes, but pd-server fails to startup if I specify more than one. I don't use TiDB layer and my clients run on a completely separate fleet and use completely separate certificates, so just do not make sense to have same CN.

### What did you do?
Specify multiple values under:
```
[security]
## A CN which must be provided by a client
cert-allowed-cn = ["TiKV", "Client"]
```
as PD config accepts multiple values under security.cert-allowed-cn: https://github.com/tikv/pd/blob/af174e6401299e0d2ac8ef72a5f4beb48be535e2/conf/config.toml#L34-L35

### What did you expect to see?
pd-server starts up and client with different CN could connect to it.

### What did you see instead?
pd-server fails to start to startup due to this check: https://github.com/tikv/pd/blob/b4c1804adbec8fe3077a0f4c29a4f6495832724c/pkg/grpcutil/grpcutil.go#L93-L103.

### What version of PD are you using (`pd-server -V`)?
5.4, but the latest master has same issue

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #5134

### What is changed and how does it work?

It can be reviewed after https://github.com/tikv/pd/pull/8441 is merged.

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

test in tiup with multi cn
- without this PR
image
- with this PR
image

### Release note

```release-note
security: support multiple CN for TLS connections
```

### CODE REVIEW

File: server/config/config.go [hunk 1]
```diff
@@ -726,20 +726,21 @@ func (c *Config) GenEmbedEtcdConfig() (*embed.Config, error) {
...
-	allowedCN, serr := c.Security.GetOneAllowedCN()
-	if serr != nil {
-		return nil, serr
-	}
 	cfg.ClientTLSInfo.ClientCertAuth = len(c.Security.CAPath) != 0
 	cfg.ClientTLSInfo.TrustedCAFile = c.Security.CAPath
 	cfg.ClientTLSInfo.CertFile = c.Security.CertPath
 	cfg.ClientTLSInfo.KeyFile = c.Security.KeyPath
-	// Client no need to set the CN. (cfg.ClientTLSInfo.AllowedCN = allowedCN)
+	// Keep compatibility with https://github.com/tikv/pd/pull/2305
```
Comment (lhy1024) [inline server/config/config.go:733]: Is it necessary?

**Comment by lhy1024** [comment]
Manual test
```
tiup cluster deploy test nightly /home/lhy1024/t.yml
tiup cluster start test --init
tiup cluster patch test /home/lhy1024/pd/bin/pd.tar.gz -R pd --overwrite
tiup cluster edit-config test
tiup cluster tls test enable
tiup cluster edit-config test
tiup cluster reload test
tiup cluster scale-out test scale-out.yaml
tiup cluster display test
mysql --comments --host 192.168.8.88 --port 4000 -u root --ssl-verify-server-cert=false -p
sysbench oltp_insert --mysql-host=192.168.8.88  --mysql-port=4000 --mysql-user=root  --mysql-password="" --tables=32 --report-interval=1 --events=0 --time=0 --threads=32 prepare
```
config file:
image

tiup display:
image

sysbench test:
image
image

pd ctl:
image


The result of sysbench is normal, tidb status display exceptions are another issue https://github.com/pingcap/tidb/issues/55372

File: client/tlsutil/tlsconfig.go [hunk 2]
```diff
@@ -121,12 +121,14 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
-	if info.AllowedCN != "" {
+	if len(info.AllowedCNs) > 0 {
 		cfg.VerifyPeerCertificate = func(_ [][]byte, verifiedChains [][]*x509.Certificate) error {
 			for _, chains := range verifiedChains {
 				if len(chains) != 0 {
-					if info.AllowedCN == chains[0].Subject.CommonName {
-						return nil
+					for _, allowedCN := range info.AllowedCNs {
+						if allowedCN == chains[0].Subject.CommonName {
+							return nil
```
Comment (rleungx) [inline client/tlsutil/tlsconfig.go:130]: directly return?

File: client/tlsutil/tlsconfig.go [hunk 3]
```diff
@@ -64,8 +64,8 @@ type TLSInfo struct {
 	// should be left nil. In that case, tls.X509KeyPair will be used.
 	parseFunc func([]byte, []byte) (tls.Certificate, error)
 
-	// AllowedCN is a CN which must be provided by a client.
-	AllowedCN string
+	// AllowedCNs is a list of CNs which must be provided by a client.
+	AllowedCNs []string
```
Comment (rleungx) [inline client/tlsutil/tlsconfig.go:68]: Can we make it unexported?

**Review by rleungx — APPROVED** [review]

File: client/tlsutil/tlsconfig.go [hunk 4]
```diff
@@ -162,8 +164,8 @@ type TLSConfig struct {
 	CertPath string `toml:"cert-path" json:"cert-path"`
 	// KeyPath is the path of file that contains X509 key in PEM format.
 	KeyPath string `toml:"key-path" json:"key-path"`
-	// CertAllowedCN is a CN which must be provided by a client
-	CertAllowedCN []string `toml:"cert-allowed-cn" json:"cert-allowed-cn"`
+	// CertAllowedCNs is the list of CN which must be provided by a client
+	CertAllowedCNs []string `toml:"cert-allowed-cn" json:"cert-allowed-cn"`
```
Comment (HuSharp) [inline client/tlsutil/tlsconfig.go:168]: This parameter doesn't be used. It seems like we can remove this parameter directly, maybe we can discuss it in another issue.

**Review by HuSharp — APPROVED** [review]
plz help to update doc as well :)

**Review by niubell — APPROVED** [review]

**Comment by lhy1024** [comment]
Manual test

## script
t.yml
```
global:
  user: "tidb"
  ssh_port: 22
  deploy_dir: "/tidb-deploy"
  data_dir: "/tidb-data"
  arch: "amd64"
server_configs:
  pd:
    replication.enable-placement-rules: true
    replication.location-labels:
      - host
monitored:
  node_exporter_port: 9100
  blackbox_exporter_port: 9115
pd_servers:
  - host: 192.168.8.88
tidb_servers:
  - host: 192.168.8.88
tikv_servers:
  - host: 192.168.8.88
    port: 20160
    status_port: 20180
    config:
      server.labels: { host: "local" }
  - host: 192.168.8.88
    port: 20161
    status_port: 20181
    config:
      server.labels: { host: "local" }
  - host: 192.168.8.88
    port: 20162
    status_port: 20182
    config:
      server.labels: { host: "local" }
tiflash_servers:
  - host: 192.168.8.88
    tcp_port: 9000
    http_port: 8123
    flash_service_port: 3930
    flash_proxy_port: 20170
    flash_proxy_status_port: 20292
    metrics_port: 8234
    config:
      server.labels: { host: "local" }
  - host: 192.168.8.88
    tcp_port: 9001
    http_port: 8124
    flash_service_port: 3931
    flash_proxy_port: 20171
    flash_proxy_status_port: 20293
    metrics_port: 8235
    config:
      server.labels: { host: "local" }
monitoring_servers:
  - host: 192.168.8.88
grafana_servers:
  - host: 192.168.8.88
alertmanager_servers:
  - host: 192.168.8.88

```

## config file
image
image
image
image

Note: we need add `test-client` because of https://github.com/pingcap/tiup/blob/a16d04fccb22106a5c8b2d7d074b7ab94a1dd677/pkg/cluster/manager/cacert.go#L65

## result

tiup display:
image

sysbench test:
image

tlflash test:
image

grafana:
image
image

**Pull request merged**